### PR TITLE
introduce elastic inflight to save memory

### LINKED ISSUE #419: Optimize memory footprint of raft state machine
After introducing [memory tracing](https://github.com/tikv/rfcs/pull/59) in TiKV, we found that the memory footprint could be very large if there were too many raft state machine. Most of them came from the inflight vector. Supposing `max_inflight_messages` was set to 4096, and there were 3 peers, leader would need to allocate 3 * 4096 * 8 = 96K memory. If there were 100k regions, then the virt memory became 9.1 GiB. And most of the memory were idle and perhaps never visited again.

There are many possible optimizations for this situation. On the one hand, if we can build a global flow control for all raft state machines, then the memory usage can be reduced; on the other hand, if we can reduce the total number of regions, the impact of flow control can also be omitted. But for the short term, I think it's easy to allocate lazily for inflight vector. We can use a linked list to allocate block by block and free them when it's idle. Because most raft state machines should be idle when it comes to 100k, so there will be only one block for all inflight vector. If each block is allocated as 512, then the memory footprint can be reduced to 572MiB, which is reasonably small enough.

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

Like #423 this PR also want to resolve #419 , but use another solution that lazy allocating inflight buffers and shrink it if necessary. No new APIs are introduced.

Would you mind to take a look? @jayzhan211

### CODE REVIEW

**Comment by jayzhan211** [comment]
LGTM

**Comment by hicqu** [comment]
@jayzhan211 thank you! I have added unit test cases for the new implementation. Could you take a look again and approve it?

**Comment by hicqu** [comment]
@jayzhan211 or could you push more commits to #423 ? Your implementation involves less allocation and free. I can compare them later.

**Comment by jayzhan211** [comment]
> @jayzhan211 or could you push more commits to #423 ? 

What commits do you mean?

File: src/tracker/inflights.rs [hunk 1]
```diff
@@ -130,127 +164,63 @@ mod tests {
...
+            assert_eq!(inflights.offset, (offset + 1) % chunk_len);
+            assert_eq!(inflights.inflights(), (1..100).collect::<Vec<_>>());
+
+            // Free to the middle item in buffer[1].
+            let to = inflights.buffer[1][3];
+            inflights.free_to(to);
+            let count = match offset {
+                0 => 80,
+                8 => 88,
+                _ => 79,
```
Comment (jayzhan211) [inline src/tracker/inflights.rs:202]: if `_` is chunk_len(15), why count is not 100 - ((16 - 15) + 4) = 95?

**Review by jayzhan211 — COMMENTED** [review]
Rest LGTM

File: src/tracker/inflights.rs (hunk 1 above)
Comment (jayzhan211) [reply src/tracker/inflights.rs:202]: > > @jayzhan211 or could you push more commits to #423 ?
> 
> What commits do you mean?


If you mean adding test, I am not sure how to test `free_*` or `reallocate_*`

**Comment by hicqu** [comment]
Here is a benchmark result:
for master branch:
```
Inflights(256)          time:   [2.1513 us 2.2001 us 2.2609 us]
Inflights(4096)         time:   [35.550 us 36.232 us 36.976 us]
```
for the patch:
```
Inflights(256)          time:   [6.8643 us 7.1716 us 7.5381 us]
                        change: [+205.14% +217.31% +228.40%] (p = 0.00 < 0.05)
                        Performance has regressed.
Inflights(4096)         time:   [103.92 us 108.21 us 113.07 us]
                        change: [+192.46% +201.84% +211.64%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

**Pull request closed without merging**